File: src/patterns/structural/adapter/broker.nim
# std
import
    std/[
        json
    ]

# internal
import
    patterns/structural/adapter/[
        governmentadapter
    ]


type
    Broker*[TAdapter: GovernmentAdapter] = ref object
        adapter: TAdapter


proc newBroker*[TAdapter: GovernmentAdapter](adapter: TAdapter): Broker[TAdapter] =
    return Broker[TAdapter](
        adapter: adapter
    )

proc sendGainsInfo*(self: Broker, data: JsonNode): bool =
    return self.adapter.sendGainsInformation(data)

File: src/patterns/structural/adapter/broker.go
package adapter


type Broker struct {
	adapter GovernmentAdapter
}


func NewBroker(adapter GovernmentAdapter) *Broker {
	return &Broker{
		adapter: adapter,
	}
}

func (self *Broker) SendGains(data []byte) bool {
	return self.adapter.SendGainsInfo(data)
}